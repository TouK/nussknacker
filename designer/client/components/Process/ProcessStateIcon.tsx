import React, {PropsWithChildren, useMemo} from "react"
import {SwitchTransition} from "react-transition-group"
import {useTranslation} from "react-i18next"
import {ProcessStateType, ProcessType} from "./types"
import {CssFade} from "../CssFade"
import {Popover} from "react-bootstrap"
import {OverlayTrigger} from "react-bootstrap/lib"
import ProcessStateUtils from "./ProcessStateUtils"
import {css, cx} from "@emotion/css"
import UrlIcon from "../UrlIcon"
import {useTheme} from "@emotion/react"
import {ProcessId} from "../../types"

interface Props {
  processState?: ProcessStateType,
  isStateLoaded?: boolean,
  process: ProcessType,
}

function Errors({state}: { state: ProcessStateType }) {
  const {t} = useTranslation()

  if (state.errors?.length < 1) {
    return null
  }

  return (
    <div>
      <span>{t("stateIcon.errors", "Errors:")}</span>
      <ul>
        {state.errors.map((error, key) => <li key={key}>{error}</li>)}
      </ul>
    </div>
  )
}

function StateIconPopover({processName, processState, tooltip, children}: PropsWithChildren<{
  processState: ProcessStateType,
  processName: ProcessId,
  tooltip: string,
}>) {
  const theme = useTheme()
  const imagePopover = useMemo(() => {
    const className = css({
      marginTop: "5px",
      "&, h3": {backgroundColor: theme.colors.primaryBackground},
    })
    return (
      <Popover id="state-icon-popover" className={className} title={processName}>
        <strong>{tooltip}</strong>
        <Errors state={processState}/>
      </Popover>
    )
  }, [processName, processState, theme.colors.primaryBackground, tooltip])

  return (
    <OverlayTrigger trigger="click" placement="left" overlay={imagePopover}>
      {children}
    </OverlayTrigger>
  )
}

function ProcessStateIcon({process, processState, isStateLoaded}: Props) {
  const icon = ProcessStateUtils.getStatusIcon(process, processState, isStateLoaded)
  const tooltip = ProcessStateUtils.getStatusTooltip(process, processState, isStateLoaded)
  // /assets/states/deploy-running-animated.svg
  const iconClass = cx(
    css({
      color: "blue",
      "--color-1": "green",
      "--color-2": "red",
    }),
    !isStateLoaded && css({
      padding: "50%",
    }),
    //TODO: normalize colors svg files
    (process.isArchived || process.isSubprocess) && css({
      filter: "invert()",
      opacity: 1,
    })
  )

  return (
    <SwitchTransition>
      <CssFade key={`${process.id}-${icon}`}>
        <StateIconPopover
          processName={process.name}
          processState={isStateLoaded ? processState : process.state}
          tooltip={tooltip}
        >
          <UrlIcon src={icon} title={tooltip} className={iconClass}/>
        </StateIconPopover>
      </CssFade>
    </SwitchTransition>
  )
}

export default ProcessStateIcon

