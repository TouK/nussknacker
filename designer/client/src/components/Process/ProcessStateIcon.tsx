import React from "react";
import { ProcessStateType, Scenario } from "./types";
import ProcessStateUtils from "./ProcessStateUtils";
import UrlIcon from "../UrlIcon";
import { Box, Divider, Popover, Typography } from "@mui/material";
import { Errors } from "./ProcessErrors";

interface Props {
    processState?: ProcessStateType;
    scenario: Scenario;
}

function ProcessStateIcon({ scenario, processState }: Props) {
    const icon = ProcessStateUtils.getStatusIcon(scenario, processState);
    const tooltip = ProcessStateUtils.getStatusTooltip(scenario, processState);
    const [anchorEl, setAnchorEl] = React.useState<HTMLButtonElement | null>(null);

    return (
        <>
            <UrlIcon src={icon} title={tooltip} onClick={(event) => processState && setAnchorEl(event.currentTarget)} />
            <Popover
                anchorEl={anchorEl}
                anchorOrigin={{
                    vertical: "center",
                    horizontal: "left",
                }}
                transformOrigin={{
                    vertical: "center",
                    horizontal: "right",
                }}
                onClose={() => setAnchorEl(null)}
                open={!!anchorEl}
            >
                <Typography p={1} variant="body1">
                    {scenario.name}
                </Typography>
                <Divider />
                <Box p={1}>
                    <Typography variant="body2" style={{ whiteSpace: "pre-wrap" }}>
                        {tooltip}
                    </Typography>
                    <Errors state={processState} />
                </Box>
            </Popover>
        </>
    );
}

export default ProcessStateIcon;
