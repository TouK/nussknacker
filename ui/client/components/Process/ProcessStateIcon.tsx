import React from "react"
import {SwitchTransition} from "react-transition-group"
import {WithTranslation, withTranslation} from "react-i18next"
import {compose} from "redux"
import {ProcessStateType, ProcessType} from "./types"
import {CssFade} from "../CssFade"

import {Popover} from "react-bootstrap"
import {OverlayTrigger} from "react-bootstrap/lib"
import ProcessStateUtils from "./ProcessStateUtils"
import {css, cx} from "@emotion/css"

type State = {
  animationTimeout: {
    enter: number,
    appear: number,
    exit: number,
  },
}

type OwnProps = {
  processState?: ProcessStateType,
  isStateLoaded: boolean,
  process: ProcessType,
  animation?: boolean,
  height?: number,
  width?: number,
  popover?: boolean,
}

type Props = OwnProps & WithTranslation

class ProcessStateIcon extends React.Component<Props, State> {
  static defaultProps = {
    isStateLoaded: false,
    processState: null,
    animation: true,
    height: 24,
    width: 24,
    popover: false,
  }

  // eslint-disable-next-line i18next/no-literal-string
  static popoverConfigs = {placement: "bottom", triggers: ["click"]}

  imageWithPopover = (image, processName: string, tooltip: string, errors: Array<string>) => {
    const {t} = this.props

    const overlay = (
      <Popover id="state-icon-popover" title={processName}>
        <strong>{tooltip}</strong>
        {errors.length !== 0 ?
          (
            <div>
              <span>{t("stateIcon.errors", "Errors:")}</span>
              <ul>
                {errors.map((error, key) => <li key={key}>{error}</li>)}
              </ul>
            </div>
          ) :
          null
        }
      </Popover>
    )

    return (
      <OverlayTrigger
        trigger={ProcessStateIcon.popoverConfigs.triggers}
        placement={ProcessStateIcon.popoverConfigs.placement}
        overlay={overlay}
      >
        {image}
      </OverlayTrigger>
    )
  }

  render() {
    const {animation, process, processState, isStateLoaded, height, width, popover} = this.props
    const icon = ProcessStateUtils.getStatusIcon(process, processState, isStateLoaded)
    const tooltip = ProcessStateUtils.getStatusTooltip(process, processState, isStateLoaded)
    const errors = (isStateLoaded ? processState?.errors : process?.state?.errors) || []

    const iconClass = cx(
      "state-icon",
      isStateLoaded === false && "state-pending",
      //TODO: normalize colors svg files
      (process.isArchived || process.isSubprocess) && css({
        filter: "invert()",
        opacity: 1,
      })
    )
    const transitionKey = `${process.id}-${icon}`

    const image = (
      <img
        src={icon}
        alt={tooltip}
        title={tooltip}
        className={iconClass}
        height={height}
        width={width}
      />
    )

    return animation === true ?
      (
        <SwitchTransition>
          <CssFade key={transitionKey}>
            {popover === true ? this.imageWithPopover(image, process.name, tooltip, errors) : image}
          </CssFade>
        </SwitchTransition>
      ) :
      image
  }
}

const enhance = compose(
  withTranslation(),
)

export default enhance(ProcessStateIcon)

