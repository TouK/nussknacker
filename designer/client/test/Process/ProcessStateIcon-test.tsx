import React from "react";
import ProcessStateIcon from "../../src/components/Process/ProcessStateIcon";
import { describe, expect, it, jest } from "@jest/globals";
import { render, waitFor } from "@testing-library/react";

const processState = {
    allowedActions: ["DEPLOY"],
    icon: "/states/stopping-success.svg",
    status: { type: "StoppedStateStatus", name: "CANCELED" },
    tooltip: "The scenario has been successfully cancelled.",
};

const noDataProcessState = {
    allowedActions: ["DEPLOY"],
    status: { type: "StoppedStateStatus", name: "CANCELED" },
};

global.fetch = jest.fn((url) =>
    Promise.resolve({
        text: () => Promise.resolve(`<svg data-testid="svg">${url}</svg>`),
    }),
) as any;

describe("ProcessStateIcon tests", () => {
    it("should show defaults for missing process.state and stateProcess", async () => {
        const scenario = { processingType: "streaming" };
        const { getByTestId, container } = render(<ProcessStateIcon scenario={scenario as any} />);
        expect(fetch).toHaveBeenLastCalledWith("/be-static/assets/states/status-unknown.svg");
        await waitFor(() => getByTestId("svg"));
        expect(container).toMatchSnapshot();
    });

    it("should show defaults for loaded process.state without data", async () => {
        const scenario = { processingType: "streaming", state: noDataProcessState };
        const { getByTestId, container } = render(<ProcessStateIcon scenario={scenario as any} />);
        await waitFor(() => getByTestId("svg"));
        expect(fetch).not.toHaveBeenCalled(); //cached
        expect(container).toMatchSnapshot();
    });

    it("should show data from loaded process.state", async () => {
        const scenario = { processingType: "streaming", state: processState };
        const { getByTestId, container } = render(<ProcessStateIcon scenario={scenario as any} />);
        expect(fetch).toHaveBeenLastCalledWith("/be-static/states/stopping-success.svg");
        await waitFor(() => getByTestId("svg"));
        expect(container).toMatchSnapshot();
    });

    it("should show defaults if loadedProcess is empty", async () => {
        const { getByTestId, container } = render(<ProcessStateIcon scenario={{} as any} />);
        await waitFor(() => getByTestId("svg"));
        expect(fetch).not.toHaveBeenCalled(); //cached
        expect(container).toMatchSnapshot();
    });

    it("should show loadedProcess data", async () => {
        const { getByTestId, container } = render(
            <ProcessStateIcon scenario={noDataProcessState as any} processState={processState as any} />,
        );
        await waitFor(() => getByTestId("svg"));
        expect(fetch).not.toHaveBeenCalled(); //cached
        expect(container).toMatchSnapshot();
    });
});
