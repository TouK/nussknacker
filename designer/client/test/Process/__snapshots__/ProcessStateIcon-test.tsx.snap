// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessStateIcon tests should show data from loaded process.state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-0"
    title="The scenario has been successfully cancelled."
  >
    <svg
      data-testid="svg"
    >
      /be-static/states/stopping-success.svg
    </svg>
  </div>
</div>
`;

exports[`ProcessStateIcon tests should show defaults for loaded process.state without data 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-0"
  >
    <svg
      data-testid="svg"
    >
      /be-static/assets/states/status-unknown.svg
    </svg>
  </div>
</div>
`;

exports[`ProcessStateIcon tests should show defaults for missing process.state and stateProcess 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-0"
  >
    <svg
      data-testid="svg"
    >
      /be-static/assets/states/status-unknown.svg
    </svg>
  </div>
</div>
`;

exports[`ProcessStateIcon tests should show defaults if loadedProcess is empty 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-0"
  >
    <svg
      data-testid="svg"
    >
      /be-static/assets/states/status-unknown.svg
    </svg>
  </div>
</div>
`;

exports[`ProcessStateIcon tests should show loadedProcess data 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-0"
    title="The scenario has been successfully cancelled."
  >
    <svg
      data-testid="svg"
    >
      /be-static/states/stopping-success.svg
    </svg>
  </div>
</div>
`;
